React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+
+    expect(getByText(title)).toBeDefined();
+  });
+});
+```
+
+### Templates test example
+
+Component:
+
+```tsx
+export const UserTemplate: FC<UserTemplateProps> = (props) => (
+  <Grid {...props}>
+    <GridItem>
+      <Navbar />
+    </GridItem>
+    <GridItem>
+      <UserCardSection title="Test title" />
+    </GridItem>
+  </Grid>
+);
+```
+
+Test:
+
+```tsx
+jest.mock("components/user-card-section");
+(UserCardSection as jest.Mock).mockReturnValue(<div />);
+
+describe("UserTemplate", () => {
+  it("Is rendering", () => {
+    const { getByTestId } = render(
+      <UserTemplate data-testid="user-template-testid" />
+    );
+
+    expect(getByTestId("user-template-testid")).toBeDefined();
+  });
+});
+```
+
+## Page component test
+
+Pages tests should be located in `src/__tests__/pages` folder because Next.js is not allowing tests in `/pages` folder. Test name should be the same as the page file with `test.tsx` extension.
+
+Page example:
+
+```tsx
+const UserPage: NextPage = () => {
+  const { data, error } = useSWR<Array<User>, IResponseError>(
+    USER_KEY,
+    fetchUsers
+  );
+
+  if (error) {
+    return <ErrorPage />;
+  }
+
+  if (!data) {
+    return <LoadingPage />;
+  }
+
+  return <UserTemplate userList={data} />;
+};
+
+export default UserPage;
+```
+
+It this page example we should test all three states of the page component. To do that we should mock `<ErrorPage>`, `<LoadingPage>` and `<UserTemplate>` and check if it's rendered based on fetcher response.
+
+Test example:
+
+```tsx
+jest.mock("components/error-page");
+jest.mock("components/loading-page");
+jest.mock("components/user-template");
+jest.mock("fetchers/users");
+
+(ErrorPage as jest.Mock).mockReturnValue(
+  <div data-testid="error-page-testid" />
+);
+(LoadingPage as jest.Mock).mockReturnValue(
+  <div data-testid="loading-page-testid" />
+);
+(UserTemplate as jest.Mock).mockReturnValue(
+  <div data-testid="user-template-page-testid" />
+);
+
+describe("User Page", () => {
+  it("Is error page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue(new Error("Error occurred!"));
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("error-page-testid")).toBeDefined();
+  });
+
+  it("Is loading page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue(null);
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("loading-page-testid")).toBeDefined();
+  });
+
+  it("Is loading page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue([{ username: "test user" }]);
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("user-template-page-testid")).toBeDefined();
+  });
+});
+```
+
+Because of testing `useSWR` behavior which will re-render DOM after fetcher promise is resolved we need to use `findByTestId` method that returns a promise that will resolve when element is added to DOM.
```

Comment:
```suggestion
Because of testing `useSWR` behavior (which will re-render DOM after fetcher promise is resolved), we need to use `findByTestId` method that returns a promise that will resolve when the element is added to DOM.
```